### PR TITLE
Integrate Stripe payment method messaging into product details page

### LINKED ISSUE #6270: Display BNPL Options on Product Details Page
### Description

As a buyer, I want to see available Buy Now Pay Later (BNPL) options on the product details page. This feature is important as it will allow me to understand my payment options before deciding to purchase.

### Acceptance criteria

1. When I am on a product detail page, I should see a section that lists available BNPL providers.
2. When I am on a product detail page, I should not see this section if there are no available BNPL providers.
3. The new section should be compatible with block based and "shortcode" based product detail pages.
### Designs

Screenshot 2023-05-11 at 16 55 46

### Dev notes

- Pending to confirm if we will use [Payment Method Messaging Element](https://href.li/?https://stripe.com/docs/payments/payment-method-messaging)
- We need to inject amount, currency, payment methods and buyer's country code into the Stripe element

### LINKED ISSUE #6503: Spike: Proof of Concept for Stripe Payment Method Messaging Integration
### Description

This issue aims to explore the feasibility and potential implementation details of integrating Stripe's Payment Method Messaging into product details page. This would involve understanding the Stripe Payment Method Messaging API, creating a proof of concept, and evaluating the impact of this integration on our current system.

### Acceptance criteria

* A proof of concept demonstrating the integration of Stripe's Payment Method Messaging into our system.

### Dev notes

[Stripe Payment Method Messaging API Documentation](https://stripe.com/docs/payments/payment-method-messaging)

### PR DESCRIPTION
**Pull request by mgascam**
Fixes #6503 

#### Changes proposed in this Pull Request
This PR integrates the Stripe's [Payment Method Messaging Element](https://stripe.com/docs/payments/payment-method-messaging). This will allow displaying information related to the available BNPL providers in the product details page.

Screenshot 2023-06-15 at 16 08 05

#### Testing instructions

_Feature flag enabled and BNPL enabled_
1. Make sure that `is_bnpl_affirm_afterpay_enabled` is enabled. You can set the `bnpl_affirm_afterpay_enabled` meta to your account in the server (don't forget to clear the cache in the client)
2. Enable Affirm and Afterpay as available payment methods in the admin ( please check instructions in this [PR](https://github.com/Automattic/woocommerce-payments-server/pull/3405) to add fees
3. Navigate to the details page of a product. Check that there's a new block below the product that displays the BNPL information (see screenshot above)
4. Increment the quantity selector. Check that the information is updated in the BNPL block

_Feature flag enabled and BNPL disabled_
1. Make sure that `is_bnpl_affirm_afterpay_enabled` is enabled. 
2. Disable Affirm and Afterpay as available payment methods in the admin
3. Navigate to the details page of a product. Check that the BNPL block is not present

_Feature flag disabled_
1. Disable feature flag for the merchant.
5. Navigate to the details page of a product. Check that the BNPL block is not present
-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6522` or branch name `dev/6503-poc-for-stripe-payment-method-messaging` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=dev/6503-poc-for-stripe-payment-method-messaging) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 8ecf0a85d95c7151cbfd28e3de90e724905e24b2
- Build time: 2023-06-16 13:56:57 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/class-wc-payments.php [hunk 1]
```diff
@@ -1633,6 +1636,19 @@ public static function init_woopay() {
 		}
 	}
 
+	/**
+	 * Load stripe site messaging script.
+	 *
+	 * @return void
+	 */
+	public static function load_stripe_site_messaging() {
+		if ( WC_Payments_Features::is_bnpl_affirm_afterpay_enabled() ) {
```
Comment (dmvrtx) [inline includes/class-wc-payments.php:1645]: Is this the right way to check for feature flag? I guess I'll have to update #6518 and even simplify it 😀

Comment (KarlisJ) [inline includes/class-wc-payments.php:1645]: Yes this introduced at some point.

**Review by KarlisJ — COMMENTED** [review]
Code changes only LGTM!

File: includes/class-wc-payments.php [hunk 2]
```diff
@@ -1633,6 +1636,19 @@ public static function init_woopay() {
...
+			require_once __DIR__ . '/class-wc-payments-payment-method-messaging-element.php';
+			$stripe_site_messaging = new WC_Payments_Payment_Method_Messaging_Element( self::$account, self::$card_gateway );
+			echo wp_kses( $stripe_site_messaging->init(), 'post' );
```
Comment (kalessil) [inline includes/class-wc-payments.php:1648]: ❤️

File: includes/class-wc-payments.php [hunk 3]
```diff
@@ -1633,6 +1636,19 @@ public static function init_woopay() {
 		}
 	}
 
+	/**
+	 * Load stripe site messaging script.
+	 *
+	 * @return void
+	 */
+	public static function load_stripe_site_messaging() {
```
Comment (kalessil) [inline includes/class-wc-payments.php]: Could we reflect somehow in naming it's BNPLs related?

File: client/product-details/index.js [hunk 4]
```diff
@@ -0,0 +1,6 @@
+/**
+ * Internal dependencies
+ */
+import siteMessaging from './site-messaging';
+
+siteMessaging();
```
Comment (kalessil) [inline client/product-details/index.js]: Cna it be converted to TS?

File: client/product-details/site-messaging/index.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
+// global Stripe, wcpayStripeSiteMessaging
+function siteMessaging() {
```
Comment (kalessil) [inline client/product-details/site-messaging/index.js]: Can it be converted to TS?

File: client/product-details/site-messaging/index.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+	// eslint-disable-next-line no-undef
+	const stripe = Stripe( publishableKey );
+	const options = {
+		amount: parseInt( price, 10 ) || 0, // In cents TODO: use quantity selector
+		currency: currency || 'USD',
+		paymentMethodTypes: paymentMethods || [],
+		countryCode: country, // Customer's country or base country of the store.
+	};
+	const elements = stripe.elements();
```
Comment (kalessil) [inline client/product-details/site-messaging/index.js]: One time-use variable?

File: includes/class-wc-payments-payment-method-messaging-element.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
+<?php
+/**
+ * Class WC_Payments_Payment_Method_Messaging_Element
+ *
+ * @package WooCommerce\Payments
+ */
+
+declare( strict_types=1 );
```
Comment (kalessil) [inline includes/class-wc-payments-payment-method-messaging-element.php]: Not sure if the minimally supported PHP version for the client supports this.

File: includes/class-wc-payments-payment-method-messaging-element.php [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+	 */
+	public function init(): string {
+		global $product;
+		$price         = $product->get_price();
+		$currency_code = get_woocommerce_currency();
+
+		if ( WC()->customer ) {
+			$billing_country = WC()->customer->get_billing_country();
+		}
+		if ( ! $billing_country ) {
```
Comment (kalessil) [inline includes/class-wc-payments-payment-method-messaging-element.php]: Perhaps 'billing_country' shall be initialized before ifs - the current implementation has the potential for accessing undefined variables.

File: includes/class-wc-payments-payment-method-messaging-element.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+			$billing_country = WC()->countries->get_base_country(); // Fallback to merchant's country.
+		}
+
+		$enabled_upe_payment_methods = $this->gateway->get_payment_method_ids_enabled_at_checkout();
+		// Filter non BNPL out of the list of payment methods.
+		$bnpl_payment_methods = array_filter(
```
Comment (kalessil) [inline includes/class-wc-payments-payment-method-messaging-element.php]: It sounds like what 'array_intersect' does.

File: includes/class-wc-payments-payment-method-messaging-element.php [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+		// register the script.
+		WC_Payments::register_script_with_dependencies( 'WCPAY_PRODUCT_DETAILS', 'dist/product-details', [ 'stripe' ] );
+		wp_enqueue_script( 'WCPAY_PRODUCT_DETAILS' );
+		// Create script tag with config.
+		wp_localize_script(
+			'WCPAY_PRODUCT_DETAILS',
+			'wcpayStripeSiteMessaging',
+			[
+				'price'          => $price * 100,
```
Comment (kalessil) [inline includes/class-wc-payments-payment-method-messaging-element.php]: With currently targeted currencies, that's ok, but we need a ticket to ensure that will work with non-decimal currencies & co as well.

File: includes/class-wc-payments-payment-method-messaging-element.php [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+				'currency'       => $currency_code,
+				'country'        => $billing_country,
+				'publishableKey' => $this->account->get_publishable_key( WC_Payments::mode()->is_test() ),
```
Comment (kalessil) [inline includes/class-wc-payments-payment-method-messaging-element.php:74]: This one is the only thing I concerned, we shall double check it's secure (if UPE exposes it somehow then we are ok)

File: client/product-details/site-messaging/index.js [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+	const paymentMessageElement = elements.create(
+		'paymentMethodMessaging',
+		options
+	);
+	paymentMessageElement.mount( '#payment-method-message' );
+
+	const quantitySelector = document.querySelector( '.quantity input' );
```
Comment (kalessil) [inline client/product-details/site-messaging/index.js]: What about variable products? E.g. T-shirt with size/colour selector?

File: includes/class-wc-payments-payment-method-messaging-element.php (hunk 10 above)
Comment (kalessil) [reply includes/class-wc-payments-payment-method-messaging-element.php]: `'price'  => WC_Payments_Utils::prepare_amount( $price,  currency_code );`

Comment (kalessil) [reply includes/class-wc-payments-payment-method-messaging-element.php]: ⬆️  should do the trick and provide a path forward for the mentioned currencies.

File: includes/class-wc-payments-payment-method-messaging-element.php (hunk 11 above)
Comment (mgascam) [reply includes/class-wc-payments-payment-method-messaging-element.php:74]: Yes, I checked and we seem to follow same approach as UPE ([ref](https://github.com/Automattic/woocommerce-payments/blob/ff4314036d3f87a16ce22d88e48af96ffc4d01e5/includes/class-wc-payments-checkout.php#L165)).

Stripe docs also confirm that it is safe to use publishablekey in client side ([ref](https://stripe.com/docs/keys)):

`On the client side: Can be publicly accessible in your web or mobile app’s client-side code (such as checkout.js) to securely collect payment information, such as with [Stripe Elements](https://stripe.com/docs/payments/elements). By default, [Stripe Checkout](https://stripe.com/docs/payments/checkout) securely collects payment information.`

File: includes/class-wc-payments.php (hunk 3 above)
Comment (mgascam) [reply includes/class-wc-payments.php]: Done here [55c6ff2](https://github.com/Automattic/woocommerce-payments/pull/6522/commits/55c6ff25ddbeb78cc19c452d3aacf33eb1fbdefd) thanks!

File: client/product-details/site-messaging/index.js (hunk 5 above)
Comment (mgascam) [reply client/product-details/site-messaging/index.js]: I'll give it a try in a following PR

File: client/product-details/index.js (hunk 4 above)
Comment (mgascam) [reply client/product-details/index.js]: I will look into this in a following PR

File: client/product-details/site-messaging/index.js (hunk 6 above)
Comment (mgascam) [reply client/product-details/site-messaging/index.js]: Addressed here [2ebca46](https://github.com/Automattic/woocommerce-payments/pull/6522/commits/2ebca46869ddf8a2077372c31d6b74d9fd031a2d) thanks!

File: includes/class-wc-payments-payment-method-messaging-element.php (hunk 7 above)
Comment (mgascam) [reply includes/class-wc-payments-payment-method-messaging-element.php]: Addressed here [55c6ff2](https://github.com/Automattic/woocommerce-payments/pull/6522/commits/55c6ff25ddbeb78cc19c452d3aacf33eb1fbdefd) thanks!

File: includes/class-wc-payments-payment-method-messaging-element.php (hunk 8 above)
Comment (mgascam) [reply includes/class-wc-payments-payment-method-messaging-element.php]: Addressed here [55c6ff2](https://github.com/Automattic/woocommerce-payments/pull/6522/commits/55c6ff25ddbeb78cc19c452d3aacf33eb1fbdefd) thanks!

File: includes/class-wc-payments-payment-method-messaging-element.php (hunk 9 above)
Comment (mgascam) [reply includes/class-wc-payments-payment-method-messaging-element.php]: Addressed here [55c6ff2](https://github.com/Automattic/woocommerce-payments/pull/6522/commits/55c6ff25ddbeb78cc19c452d3aacf33eb1fbdefd) thanks!

File: includes/class-wc-payments-payment-method-messaging-element.php (hunk 10 above)
Comment (mgascam) [reply includes/class-wc-payments-payment-method-messaging-element.php]: Addressed here [55c6ff2](https://github.com/Automattic/woocommerce-payments/pull/6522/commits/55c6ff25ddbeb78cc19c452d3aacf33eb1fbdefd) thanks!

File: client/product-details/site-messaging/index.js (hunk 12 above)
Comment (mgascam) [reply client/product-details/site-messaging/index.js]: I'll take care of this in a following PR

**Review by kalessil — APPROVED** [review]
LGTM!

File: includes/class-wc-payments-payment-method-messaging-element.php [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+		$billing_country = WC()->countries->get_base_country();
+
+		if ( WC()->customer ) {
+			$billing_country = WC()->customer->get_billing_country(); // Use the customer's billing country if available.
+		}
+
+		$enabled_upe_payment_methods = $this->gateway->get_payment_method_ids_enabled_at_checkout();
+		// Filter non BNPL out of the list of payment methods.
+
+		// Filter out non-BNPL payment methods.
```
Comment (kalessil) [inline includes/class-wc-payments-payment-method-messaging-element.php]: Just a small remark - this comment seems to be the same as the previous one.

Comment (mgascam) [reply includes/class-wc-payments-payment-method-messaging-element.php]: Oops, removed here [8ecf0a8](https://github.com/Automattic/woocommerce-payments/pull/6522/commits/8ecf0a85d95c7151cbfd28e3de90e724905e24b2) thanks!

**Pull request merged**